crypto/evp/pmeth_gn.c

```diff
@@ -150,23 +150,53 @@ int EVP_PKEY_CTX_get_keygen_info(EVP_PKEY_CTX *ctx, int idx)
     return ctx->keygen_info[idx];
 }
 
+static EVP_PKEY *evp_pkey_new_key(int priv, int type, ENGINE *e,
+                                  const unsigned char *key, int keylen)
+{
+    EVP_PKEY_CTX *ctx = NULL;
+    EVP_PKEY *pkey = NULL;
+
+    /* Note: for compatibility reasons we allow keylen to be < 0 in some cases */
```

Comment:
Could you please expand that? What compatibility reasons?

Reply:
Basically the old EVP_PKEY_new_mac_key() has this weird behaviour where, for HMAC only, if you set a keylen of -1 then it assumes that the (binary!) key is NUL terminated! We use that capability in the dgst app. In refreshing my memory on why this was the case I realised I went too far in the dgst app renaming EVP_PKEY_new_mac_key() to EVP_PKEY_new_priv_key(). It should stay as EVP_PKEY_new_mac_key() because of this oddity (the new functions take a size_t for the length so this isn't possible with them). I expanded out the comment to explain this better, and fixed the dgst app.